### ISSUE TITLE
Not able to read .ply voxel output.

### ISSUE THREAD

**Issue by AzamatB** [mkazhdan/PoissonRecon#28]
```
 PoissonRecon --in bunny.points.ply --voxel bunny.voxel.ply --depth 10
```

My understanding that the line above outputs 3D array (voxel grid), where each element is a value of the implicit function.

**Comment by mkazhdan** [comment]
That's right. Though the first 4 bytes store the binary encoding of the resolution of the grid (2^d).

-- misha

**Comment by AzamatB** [comment]
``` C++
#include <iostream>
#include <pcl/io/pcd_io.h>
#include <pcl/point_types.h>
#include <pcl/features/normal_3d.h>
#include <pcl/io/ply_io.h>

int main (int argc, char** argv) {
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZ>);

    if (pcl::io::loadPLYFile<pcl::PointXYZ> ("bunny.voxel.ply", *cloud) == -1) //* load the file
    {
        PCL_ERROR ("Couldn't read file test_pcd.pcd \n");
        return (-1);
    }
    std::cout << "Loaded "
              << cloud->width * cloud->height
              << " data points from test_pcd.pcd with the following fields: "
              << std::endl;
    return (0);
}

```

**Comment by mkazhdan** [comment]
I'm not sure I follow the question here. Are you trying to read the voxel grid as a ply file? That won't work. It's written in raw binary, so that if r=2^d is the resolution, the file will have size (r*r*r*+1)*4 bytes. With the first 4 bytes used to store the binary representation of the integer r and the remaining r*r*r*4 bytes storing the binary representation of r*r*r single-precision floating point values​ of the implicit function.

**Comment by AzamatB** [comment]
Ok, I'll try to explain.
My goal is to get the 3D binary image of the object, i.e. voxelized representation instead of meshes.
To do this I was planning to read 3D tensor of single-precision floating point values​ of the implicit function that as I understand is stored in the `bunny.voxel.ply` file produced by
`PoissonRecon --in bunny.points.ply --voxel bunny.voxel.ply --depth 10`
and then work with it to get the binary 3D tensor from it.
But I cannot retrieve (i.e. read into the memory) the 3D tensor of the implicit function values from `bunny.voxel.ply`.

**Comment by mkazhdan** [comment]
Assume you just want the 3D array of values, you can read it as:

FILE* fp = fopen( "bunny.voxel" , "rb" );
int res;
fread( &res , 1 , sizeof(int) , fp );
float *voxel = new float[res*res*res];
fread( voxel , res*res*res , sizeof(float) , fp );
fclose( fp );

**Comment by AzamatB** [comment]
Great! Thank you!
One clarification: in this code `voxel` seems to be 1D array.
So say if I want to access element, which in voxel grid (3D tensor), has indicies [x][y][z].
how can I access it from `voxel`

**Comment by mkazhdan** [comment]
voxel(x,y,z)=voxel[res*res*z+res*y+x]

**Comment by mikeroberts3000** [comment]
On a related note, does the volume always extend spatially from (0,0,0) to (1,1,1)? This question is related to https://github.com/mkazhdan/PoissonRecon/issues/21.

**Comment by mkazhdan** [comment]
Yep.

**Comment by wangmiaowei** [comment]
I write a python code to load this voxel:
import numpy as np
```
def load_voxel_grid(file_name):
    with open(file_name, 'rb') as f:

        res = np.frombuffer(f.read(4), dtype=np.int32)[0]
        print(f"Resolution: {res}")

     
        resolution = res ** 3
        
     
        voxel_values = np.frombuffer(f.read(resolution * 4), dtype=np.float32)
        
    return res, voxel_values

if __name__ == '__main__':
    file_name = 'real_mesh.voxel' 
    resolution, voxel_values = load_voxel_grid(file_name)
    

    print("Voxel Values:", voxel_values[:10])
```
but the resolution is very large: Resolution: 822752071 any issue?

**Comment by mkazhdan** [comment]
I'm afraid I don't understand the question or it's relation to the Poisson Reconstruction code.

**Comment by mkazhdan** [comment]
Apologies. The grid format was changed a while back and the description in the readme was out of sync.
I have modified it so that the description of the contents now matches what is output.

**Comment by wangmiaowei** [comment]
yes, there is an updation.
